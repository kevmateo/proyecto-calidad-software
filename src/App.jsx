import './App.css';
import { Form, InputGroup } from 'react-bootstrap';
import TarjetaPalabra from './componentes/tarjeta-palabra';
import LoadingPage from './componentes/loading-page';
import { useState, useEffect, Suspense } from 'react';
import NavBar from './componentes/navbar.jsx';
import { CgDanger } from "react-icons/cg";
import datos_quemados from './datos_quemados.json';

function App(props) {

  const [palabras, setPalabras] = useState([]);
  const [palabrasAuxiliares, setPalabrasAuxiliares] = useState([]);
  const [filtroNombre, setFiltroNombre] = useState("");
  const [isLoading, setIsLoading] = useState(true);
  const [isSearching, setIsSearching] = useState(false);
  const [isVisible, setIsVisible] = useState(false);
  const [ultimoResultado, setUltimoResultado] = useState([]);
  const [darkMode, setDarkMode] = useState(false);

  useEffect(() => {
    if (props.setIsLoading === false) {
      setIsLoading(false); // Cambia isLoading a false cuando el componente se monta
    }

  }, [props.setIsLoading]);

  useEffect(() => {
    if (palabrasAuxiliares.length > 0) {
      setUltimoResultado(palabrasAuxiliares);
    }
  }, [palabrasAuxiliares]);

  useEffect(() => {
    if (palabrasAuxiliares.length === 0 && ultimoResultado.length > 0) {
      const timeoutId = setTimeout(() => {
        setUltimoResultado([]);
      }, 300);

      return () => clearTimeout(timeoutId);
    } else if (palabrasAuxiliares.length > 0) {
      setUltimoResultado(palabrasAuxiliares);
    }
  }, [palabrasAuxiliares]);

  
  const handlerTraerPalabras = () => {
    fetch(`http://3.86.140.231/diccionary/imagenes`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    })
      .then((response) => {
        if (response.ok) {
          return response.json();
        } else {
          console.log("Error al obtener los platos");
        }
      })
      .then((data) => {
        //localStorage.setItem("imagen", JSON.stringify(data));
        setPalabras(data);
        setPalabrasAuxiliares(data);
        console.log(data);
      })
      .catch((error) => {
        console.log("Error de red: " + error);
      });
  };

  useEffect(() => {
    // const imagenesGuardadas = localStorage.getItem("imagen");
    // if (imagenesGuardadas) {
    //   const imagen = JSON.parse(imagenesGuardadas);
    //   setPalabras(imagen);
    //   setPalabrasAuxiliares(imagen);
    //   console.log(imagen);
    // } else {
    handlerTraerPalabras();
    //}
  }, []);

  const handlerBuscarPalabras = (e) => {
    const texto = e.target.value;
    setFiltroNombre(texto);
    const cleanText = texto.trim();
  
    const palabrasFiltradas = palabras.filter((p) => {
      const filterBySansckrit =
        p.sansckrit && p.sansckrit.toLowerCase().includes(cleanText.toLowerCase());
      const filterByEnglish =
        p.english && p.english.toLowerCase().includes(cleanText.toLowerCase());
      const filterBySpanish =
        p.spanish && p.spanish.toLowerCase().includes(cleanText.toLowerCase());
  
      const filterByWordModels = p.diccionaryWordModels.some((wp) => {
        const spanishWord =
          wp.word.spanish_w &&
          wp.word.spanish_w.toLowerCase().includes(cleanText.toLowerCase());
        const englishWord =
          wp.word.english_w &&
          wp.word.english_w.toLowerCase().includes(cleanText.toLowerCase());
        const sansWord =
          wp.word.sans_word &&
          wp.word.sans_word.toLowerCase().includes(cleanText.toLowerCase());
        return spanishWord || englishWord || sansWord;
      });
      return (
        filterBySansckrit || filterByEnglish || filterBySpanish || filterByWordModels
      );
    });
    setPalabrasAuxiliares(palabrasFiltradas);
  };

  const handleInputFocus = () => {
    setIsSearching(true);
  };

  const handleAnimationEnd = () => {
    setTimeout(() => {
      setIsLoading(false);
    }, 1500);

    setTimeout(() => {
      setIsVisible(true);
    }, 1600);
  };

  const handlerToggleDarkMode = () => {
    setDarkMode(!darkMode);
    const body = document.body;
    if (darkMode) {
      body.classList.remove('dark-theme');
    } else {
      body.classList.add('dark-theme');
    }
  }

  return (
    <div className='contenido-principal-1'>
      {isLoading && <LoadingPage handleAnimationEnd={handleAnimationEnd} />}
      {!isLoading && (
        <>
          <NavBar id="nav-bar" darkMode={darkMode} toggleDarkMode={handlerToggleDarkMode} />
          <div className={`contenedor-principal ${isVisible ? 'visible' : ''}`}>
            <div className='contenedor-titulo'>
              <h1>SansGoogle</h1>
              <InputGroup className="input-gruop">
                <Form.Control data-bs-theme={darkMode ? 'dark' : 'light'}
                  placeholder="Escriba para empezar..."
                  aria-label="Recipient's username"
                  aria-describedby="basic-addon2"
                  value={filtroNombre}
                  onChange={handlerBuscarPalabras}
                  onFocus={handleInputFocus}
                />
              </InputGroup>
            </div>
            <Suspense fallback={<h2>Buscando...</h2>}>
              <div className={`contenedor-tarjetas ${(isSearching && palabrasAuxiliares.length > 0) ? 'visible' : ''}`}>
                {ultimoResultado.map((palabra, index) => (
                  <TarjetaPalabra darkMode={darkMode}
                    key={index}
                    id_s={palabra.id_s}
                    sansckrit={palabra.sansckrit}
                    english={palabra.english}
                    spanish={palabra.spanish}
                    diccionaryWordModels={palabra.diccionaryWordModels}
                    link={palabra.link}
                  />
                ))}
              </div>
            </Suspense>
            {isSearching && palabrasAuxiliares.length === 0 && (
              <div className="contenedor-sin-resultados">
                <p ><CgDanger size={20}/> No se encontraron resultados</p>
              </div>
            )}
          </div>
        </>
      )}
    </div>
  );
}

export default App;
