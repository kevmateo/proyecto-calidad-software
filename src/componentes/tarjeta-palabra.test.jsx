import React from "react";
import '@testing-library/jest-dom/extend-expect';
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import TarjetaPalabra from "./tarjeta-palabra.jsx";
import App from '../App.jsx';
import { Selector, fixture } from 'testcafe';


const palabra = {
  id_palabra: 1,
  sansckrit: 'Tadasana',
  spanish: 'Postura de la Montaña',
  english: 'Mountain Pose',
  diccionaryWordModels: [
    {
      id_diccionary_word: {
        id_s: 1,
        id_w: 1,
      },
      word: {
        id: 1,
        sans_word: 'Asana',
        spanish_w: 'postura',
        english_w: 'pose',
      },
    },
    {
      id_diccionary_word: {
        id_s: 1,
        id_w: 25,
      },
      word: {
        id: 25,
        sans_word: 'Tad',
        spanish_w: 'montaña',
        english_w: 'mountain',
      },
    },
  ],
};

test('renderiza el componente correctamente', () => {
  const ejemploProps = {
    spanish: 'Postura de la Montaña',
    english: 'Mountain Pose',
    sansckrit: 'Tadasana',
  };

  const componente = render(<TarjetaPalabra ejemploProps={ejemploProps} />);

  expect(componente).toMatchSnapshot();
});

test('click en el boton de cambiar idioma', () => {
  const ejemploProps = {
    spanish: 'Postura de la Montaña',
    english: 'Mountain Pose',
    sansckrit: 'Tadasana',
  };

  const componente = render(<TarjetaPalabra ejemploProps={ejemploProps} />);

  const botonSanscrito = componente.getByRole('button', { name: 'Sánscrito' });
  const botonEnglish = componente.getByRole('button', { name: 'Inglés' });
  const botonSpanish = componente.getByRole('button', { name: 'Español' });

  fireEvent.click(botonSanscrito);
  fireEvent.click(botonEnglish);
  fireEvent.click(botonSpanish);
});

test('verificar la existencia de elementos en el componente', () => {
  const ejemploProps = {
    spanish: 'Postura de la Montaña',
    english: 'Mountain Pose',
    sansckrit: 'Tadasana',
  };

  const componente = render(<TarjetaPalabra ejemploProps={ejemploProps} />);

  const spanishButton = componente.getByRole('button', { name: 'Español' });
  const englishButton = componente.getByRole('button', { name: 'Inglés' });
  const sansckritButton = componente.getByRole('button', { name: 'Sánscrito' });

  expect(componente).toMatchSnapshot();
  expect(spanishButton).toBeInTheDocument();
  expect(englishButton).toBeInTheDocument();
  expect(sansckritButton).toBeInTheDocument();
});

describe('TarjetaPalabra Component', () => {

  test('debería mostrar la palabra en el idioma seleccionado', () => {
    render(<TarjetaPalabra {...palabra} />);

    expect(screen.getByText('Tadasana')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Español'));
    expect(screen.getByText('Postura de la Montaña')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Inglés'));
    expect(screen.getByText('Mountain Pose')).toBeInTheDocument();
  });

  test('debería mostrar la descomposición morfológica con la palabra', () => {
    render(<TarjetaPalabra {...palabra} />);

    expect(screen.getByText('Asana')).toBeInTheDocument();
    expect(screen.getByText('Tad')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Español'));
    expect(screen.getByText('postura')).toBeInTheDocument();
    expect(screen.getByText('montaña')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Inglés'));
    expect(screen.getByText('pose')).toBeInTheDocument();
    expect(screen.getByText('mountain')).toBeInTheDocument();
  });
});

test('debería mostrar la palabra buscada con su descomposición morfológica y opciones de idioma', async () => {
  render(<App setIsLoading={false} />);

  const inputBusqueda = screen.getByLabelText("Recipient's username");
  fireEvent.change(inputBusqueda, { target: { value: 'Tadasana' } });

  await waitFor(() => {
    const tadasanaElements = screen.getAllByText('Tadasana');
    expect(tadasanaElements.length).toBeGreaterThan(0);

    const asanaElements = screen.getAllByText('Asana');
    expect(asanaElements.length).toBeGreaterThan(0);

    const spanishButtons = screen.getAllByText('Español');
    expect(spanishButtons.length).toBeGreaterThan(0);

    const englishButtons = screen.getAllByText('Inglés');
    expect(englishButtons.length).toBeGreaterThan(0);
  });
});

test('Renderiza la tarjeta con la palabra y sus traducciones en diferentes idiomas', () => {
  render(<TarjetaPalabra
    spanish={palabra.spanish}
    sansckrit={palabra.sansckrit}
    english={palabra.english}
    diccionaryWordModels={palabra.diccionaryWordModels}
  />);

  const sansckritWordElement = screen.getByText('Tadasana');
  expect(sansckritWordElement).toBeInTheDocument();

  const buttonEspanol = screen.getByText('Español');
  fireEvent.click(buttonEspanol);
  const spanishWordElement = screen.getByText('Postura de la Montaña');
  expect(spanishWordElement).toBeInTheDocument();

  const buttonIngles = screen.getByText('Inglés');
  fireEvent.click(buttonIngles);
  const englishWordElement = screen.getByText('Mountain Pose');
  expect(englishWordElement).toBeInTheDocument();
});

test('La traducción de las palabras cambia según el idioma seleccionado en menos de 3 segundos', async () => {
  render(<TarjetaPalabra {...palabra} />);

  expect(screen.getByText('Tadasana')).toBeInTheDocument();
  expect(screen.getByText('Asana')).toBeInTheDocument();

  fireEvent.click(screen.getByText('Español'));
  await waitFor(() => {
    expect(screen.getByText('Postura de la Montaña')).toBeInTheDocument();
    expect(screen.getByText('postura')).toBeInTheDocument();
  }, { timeout: 3000 });


  fireEvent.click(screen.getByText('Inglés'));
  await waitFor(() => {
    expect(screen.getByText('Mountain Pose')).toBeInTheDocument();
    expect(screen.getByText('pose')).toBeInTheDocument();
  }, { timeout: 3000 });
});

fixture`App Navegadores Compatibles`.page`http://localhost:3000`;
test('Verificar compatibilidad con distintos navegadores', async (t) => {

  await t.expect(Selector('h1').innerText).eql('SansGoogle');

  await t.expect(Selector('.input-gruop').visible).ok();
  await t.expect(Selector('.contenedor-tarjetas').visible).ok();

  await t
    .useRole(chromeUser)
    .expect(Selector('h1').innerText)
    .eql('SansGoogle')
    .expect(Selector('.input-gruop').visible)
    .ok()
    .expect(Selector('.contenedor-tarjetas').visible)
    .ok();

  await t
    .useRole(firefoxUser)
    .expect(Selector('h1').innerText)
    .eql('SansGoogle')
    .expect(Selector('.input-gruop').visible)
    .ok()
    .expect(Selector('.contenedor-tarjetas').visible)
    .ok();

});