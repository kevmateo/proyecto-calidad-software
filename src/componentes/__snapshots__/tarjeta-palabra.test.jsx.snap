// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderiza el componente correctamente 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card"
        data-bs-theme="light"
        style="width: 18rem; height: 550px;"
      >
        <img
          class="card-img-top"
          src="data:image/jpeg;base64,undefined"
          style="height: 200px;"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title h5"
          />
        </div>
        <div
          class="list-group-flush list-group"
        />
        <div
          class="card-body-botones card-body"
        >
          <button
            class="boton-sanscrito btn btn-primary"
            type="button"
          >
            Sánscrito
          </button>
          <button
            class="boton-español btn btn-primary"
            type="button"
          >
            Español
          </button>
          <button
            class="boton-ingles btn btn-primary"
            type="button"
          >
            Inglés
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card"
      data-bs-theme="light"
      style="width: 18rem; height: 550px;"
    >
      <img
        class="card-img-top"
        src="data:image/jpeg;base64,undefined"
        style="height: 200px;"
      />
      <div
        class="card-body"
      >
        <div
          class="card-title h5"
        />
      </div>
      <div
        class="list-group-flush list-group"
      />
      <div
        class="card-body-botones card-body"
      >
        <button
          class="boton-sanscrito btn btn-primary"
          type="button"
        >
          Sánscrito
        </button>
        <button
          class="boton-español btn btn-primary"
          type="button"
        >
          Español
        </button>
        <button
          class="boton-ingles btn btn-primary"
          type="button"
        >
          Inglés
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`verificar la existencia de elementos en el componente 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card"
        data-bs-theme="light"
        style="width: 18rem; height: 550px;"
      >
        <img
          class="card-img-top"
          src="data:image/jpeg;base64,undefined"
          style="height: 200px;"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title h5"
          />
        </div>
        <div
          class="list-group-flush list-group"
        />
        <div
          class="card-body-botones card-body"
        >
          <button
            class="boton-sanscrito btn btn-primary"
            type="button"
          >
            Sánscrito
          </button>
          <button
            class="boton-español btn btn-primary"
            type="button"
          >
            Español
          </button>
          <button
            class="boton-ingles btn btn-primary"
            type="button"
          >
            Inglés
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card"
      data-bs-theme="light"
      style="width: 18rem; height: 550px;"
    >
      <img
        class="card-img-top"
        src="data:image/jpeg;base64,undefined"
        style="height: 200px;"
      />
      <div
        class="card-body"
      >
        <div
          class="card-title h5"
        />
      </div>
      <div
        class="list-group-flush list-group"
      />
      <div
        class="card-body-botones card-body"
      >
        <button
          class="boton-sanscrito btn btn-primary"
          type="button"
        >
          Sánscrito
        </button>
        <button
          class="boton-español btn btn-primary"
          type="button"
        >
          Español
        </button>
        <button
          class="boton-ingles btn btn-primary"
          type="button"
        >
          Inglés
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
